core/src/main/java/com/sequenceiq/cloudbreak/service/NodeCountLimitValidator.java

```diff
@@ -0,0 +1,43 @@
+package com.sequenceiq.cloudbreak.service;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.inject.Inject;
+
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.base.InstanceGroupV4Base;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.request.StackV4Request;
+import com.sequenceiq.cloudbreak.common.exception.BadRequestException;
+import com.sequenceiq.cloudbreak.conf.LimitConfiguration;
+import com.sequenceiq.cloudbreak.service.stack.InstanceMetaDataService;
+
+@Service
+public class NodeCountLimitValidator {
+
+    @Inject
+    private InstanceMetaDataService instanceMetaDataService;
+
+    @Inject
+    private LimitConfiguration nodeCountLimitConfiguration;
+
+    public void validateScale(Long stackId, Integer scalingAdjustment) {
+        if (scalingAdjustment > 0) {
+            Integer currentNodeCount = instanceMetaDataService.countByStackId(stackId).getInstanceCount();
+            validateNodeCount(currentNodeCount + scalingAdjustment);
+        }
+    }
+
+    public void validateProvision(StackV4Request stackRequest) {
+        AtomicInteger nodeCounter = new AtomicInteger(0);
+        stackRequest.getInstanceGroups().stream().map(InstanceGroupV4Base::getNodeCount).forEach(igNodeCount -> nodeCounter.addAndGet(igNodeCount));
```

Comment:
I think it could be simpler, for example:
`Integer nodeCount = stackRequest.getInstanceGroups().stream().mapToInt(InstanceGroupV4Base::getNodeCount).sum();`